src/auth/executableresponse.ts

```diff
@@ -0,0 +1,166 @@
+ * pluggable auth client.
+ */
+export interface ExecutableResponseJson {
+  /**
+   * Version of the response, only version 1 currently supported.
+   * Always required.
+   */
+  version: number;
+  /**
+   * Whether the executable ran successfully. Always required.
+   */
+  success: boolean;
+  /**
+   * Epoch time for expiration of the token in seconds, required for successful
+   * responses.
+   */
+  expiration_time?: number;
+  /**
+   * Type of subject token in the response, currently supported values are:
+   * urn:ietf:params:oauth:token-type:saml2
+   * urn:ietf:params:oauth:token-type:id_token
+   * urn:ietf:params:oauth:token-type:jwt
+   */
+  token_type?: string;
+  /**
+   * Error code from executable, required when unsuccessful.
+   */
+  code?: string;
+  /**
+   * Error message from executable, required when unsuccessful.
+   */
+  message?: string;
+  /**
+   * ID token to be used as a subject token when token_type is id_token or jwt.
+   */
+  id_token?: string;
+  /**
+   * Response to be used as a subject token when token_type is saml2.
+   */
+  saml_response?: string;
+}
+
+/**
+ * Class defining a response from a 3rd party executable run by the pluggable
+ * auth client.
+ */
+export class ExecutableResponse {
+  readonly version: number;
+  readonly success: boolean;
+  readonly expirationTime?: number;
+  readonly tokenType?: string;
+  readonly errorCode?: string;
+  readonly errorMessage?: string;
+  readonly subjectToken?: string;
+
+  /**
+   * Instantiates an ExecutableResponse instance using the provided JSON object
+   * from the output of the executable.
+   * @param responseJson Response from a 3rd party executable, loaded from a
+   * run of the executable or a cached output file.
+   */
+  constructor(responseJson: ExecutableResponseJson) {
+    // Check that always required fields exist in the json response.
+    if (!responseJson.version) {
+      throw Error("Executable response must contain a 'version' field.");
+    }
+    if (responseJson.success === undefined) {
+      throw Error("Executable response must contain a 'success' field.");
+    }
+
+    this.version = responseJson.version;
+    this.success = responseJson.success;
+
+    // Check that fields required when response is successful exist.
+    if (this.success) {
+      if (!responseJson.expiration_time) {
+        throw Error(
+          "Executable response must contain an 'expiration_time' field when successful."
+        );
+      }
+      if (!responseJson.token_type) {
+        throw Error(
+          "Executable response must contain a 'token_type' field when successful."
+        );
+      }
+      this.expirationTime = responseJson.expiration_time;
+      this.tokenType = responseJson.token_type;
+
+      // Check that type specific fields exist based on token_type provided and
+      // set subjectToken.
+      if (this.tokenType === SAML_SUBJECT_TOKEN_TYPE) {
```

Comment:
Maybe we can split this up to reduce nesting and make it easier to follow:

```
e.g.:

if (
  this.tokenType !== SAML_SUBJECT_TOKEN_TYPE1 ||
  this.tokenType !== OIDC_SUBJECT_TOKEN_TYPE1 ||
  this.tokenType !== OIDC_SUBJECT_TOKEN_TYPE2
) {
 throw Error(...);
} 

if (this.tokenType === SAML_SUBJECT_TOKEN_TYPE) {
  this.subjectToken = responseJson.saml_response;
} else {
  this.subjectToken = responseJson.id_token;
}

if (!this.subjectToken) throw...
```

Reply:
You can still do that - just add an error for each. Comment still applies. Also the error doesn't have to be generic either, you can put 'saml_response' or 'id_token' but might as well have specific ones